design-proposals/klip-28-create-or-replace

```diff
@@ -0,0 +1,156 @@
+# KLIP 28 - Introduce CREATE OR REPLACE
+
+**Author**: agavra | 
+**Release Target**: 0.11 | 
+**Status**: _In Discussion_ | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** _CREATE OR REPLACE is a mechanism geared toward enabling in-place ksqlDB query evolution._
+           
+## Motivation and background
+
+Production deployments of databases are never static; they evolve as application and business 
+requirements change. To that end, all popular data stores have ways of managing and manipulating 
+existing data. For stream processing applications, a user may want to modify their application as 
+a result of:
+
+- Business Requirements: requirements simply change over time
+- Schema Evolution: the incoming data or required output has been modified
+- Optimizations: the same application can be executed more efficiently (either by user or engine)
+
+At time of writing, ksqlDB provides a crude mechanism for altering its application behavior: dropping 
+a query and restarting it at the earliest or latest offset. While this often works well in development 
+environments, there are limitations to its applicability in production:
+
+- Data Retention: the earliest available offset may not correspond with the beginning of time
+- Downtime: the delta between terminating and catch-up may be out of the application’s SLO
+- Output Routing: populating to the old output topic will produce duplicates while using a new output topic will require cascading migrations
+- Compute: recomputing the complete history for a query may not be feasible
+
+Kafka Streams provides more granular mechanisms (e.g. restarting queries with different behaviors 
+but identical consumer groups), but these methods burden users with extra complexity and lack guardrails.
```

Comment:
Instead of 

> but these methods burden users with extra complexity and lack guardrails

how about

> but to directly expose these methods to users through ksqlDB would burden them with extra complexity and lack guardrails

Or, perhaps I'm misunderstanding the intention here?

Reply:
I think both statements are true - the intention in this KLIP is also to call out that there are improvements that need to be made in Kafka Streams to ensure that an upgrade is valid. Some of that burden will be on KSQL in the short term, but the intention is to address some of these concerns inside Kafka Streams as well